app/javascript/src/ActiveDocs/ThreeScaleApiDocs.ts

```diff
@@ -17,6 +20,96 @@ const appendSwaggerDiv = (container: HTMLElement, id: string): void => {
   container.appendChild(div)
 }
 
+/**
+ * when using Record notation, the following error is thrown:
+ * 'TS2456: Type alias 'BodyValue' circularly references itself.'
+ */
+// eslint-disable-next-line @typescript-eslint/consistent-indexed-object-style
+type BodyValue = boolean | number | string | { [key: string]: BodyValue }
+
+/**
+ * Transforms an object into form data representation, also URL-encoding the values,
+ * Example:
+ * {
+ *   a_string: 'hello',
+ *   an_array: [
+ *     { first: 1 },
+ *     { second: 1, extra_param: 'with whitespace'}
+ *   ]
+ * }
+ * =>
+ * {
+ *   a_string: 'hello',
+ *   'an_array[0][first]': '1',
+ *   'an_array[1][second]': '1',
+ *   'an_array[1][extra_param]': 'with%20whitespace'
+ * }
+ * @param object
+ */
+export const objectToFormData = (object: BodyValue): Record<string, boolean | number | string> => {
+  const buildFormData = (formData: Record<string, boolean | number | string>, data: BodyValue, parentKey?: string) => {
+    if (data && typeof data === 'object') {
+      Object.keys(data).forEach((key: string) => {
+        buildFormData(formData, data[key], parentKey ? `${parentKey}[${key}]` : key)
+      })
+    } else {
+      if (parentKey) {
+        formData[parentKey] = data ? data : ''
+      }
+    }
+  }
+  const formData = {}
+  buildFormData(formData, object)
+  return formData
+}
+
+/**
+ * Transforms the request body of the Service Management API Report, as
+ * SwaggerUI (or rather swagger-js) does not serialize arrays of objects properly
+ * The hack is to process the request body in two steps:
+ * 1. normalize the 'transactions' array, as its elements may be either an object (if the value is taken
+ *    from the example in the spec), or as a serialized JSON (if the field is changed manually)
+ * 2. "flatten" the objects by transforming them into form-data structure with the entries like
+ *    'transactions[0][app_id]': 'example'
+ *    'transactions[0][usage][hits]': 1
+ * @param body BackendApiReportBody
+ */
+export const transformReportRequestBody = (body: BackendApiReportBody): Record<string, boolean | number | string> => {
+  if (Array.isArray(body.transactions)) {
+    body.transactions = body.transactions.map(transaction => {
+      switch (typeof transaction) {
+        case 'object':
+          return transaction
+        case 'string':
+          try {
+            return JSON.parse(transaction) as BackendApiTransaction
+          } catch (error: unknown) {
+            return null
+          }
```

Comment:
`safeFromJsonString` catches errors and casts types:
```suggestion
          return safeFromJsonString<BackendApiTransaction>(transaction) ?? null
```
A question though, instead of filtering out `null` attributes, could we set `undefined` as default and it will be ignored without the need of a filter? Example:
```js
JSON.stringify({ a: undefined, b: 123 })
// -> "{\"b\":123}" 
```

Reply:
Nice fact about `JSON.stringify` trimming the undefined attributes...
But in this case we want an object, not a stringified JSON value. And I don't think stringifying and then parsing again is reasonable for the purpose of getting rid of empty values.

As an alternative, instead of the `.map().filter()` chain I can just do a simple loop populating a new array with non-empty objects only, maybe that would be less confusing and more efficient.

Reply:
Removed the filtering and added `safeFromJsonString` in this commit https://github.com/3scale/porta/pull/3294/commits/5bd22380662cf65087c92006144ec8368c0d59a3#diff-78d4b14487564cb474b2638642cd81f8cbc2626033cfe769bb0fe1f6524af4f6R77